controller/object_store_controller.go

```diff
@@ -0,0 +1,1190 @@
+// needed to operate the object stores. If any of them is found to be
+// unhealthy, the controller will transition the object store to "Error"
+// state, otherwise do nothing.
+func (osc *ObjectStoreController) handleRunning(store *longhorn.ObjectStore) (err error) {
+	if store.Spec.TargetState == longhorn.ObjectStoreStateStopped {
+		logrus.Infof("object store %v is now stopping", store.Name)
+		store.Status.State = longhorn.ObjectStoreStateStopping
+		return nil
+	}
+
+	dpl, err := osc.ds.GetDeployment(store.Name)
+	if err != nil && datastore.ErrorIsNotFound(err) {
+		return errors.Wrapf(err, "failed to find deployment %v", store.Name)
+	} else if err = osc.checkDeployment(dpl, store); err != nil {
+		logrus.Errorf("Object Store running but deployment not ready")
+		return err
+	}
+
+	_, err = osc.ds.GetService(osc.namespace, store.Name)
+	if err != nil && datastore.ErrorIsNotFound(err) {
+		return errors.Wrapf(err, "failed to find service %v", store.Name)
+	}
+
+	vol, err := osc.ds.GetVolume(genPVName(store))
+	if err != nil && datastore.ErrorIsNotFound(err) {
+		return errors.Wrapf(err, "failed to find volume %v", genPVName(store))
+	} else if err = osc.checkVolume(vol); err != nil {
+		logrus.Errorf("Object Store running but Volume not ready")
+		return err
+	}
+
+	pv, err := osc.ds.GetPersistentVolume(genPVName(store))
+	if err != nil && datastore.ErrorIsNotFound(err) {
+		return errors.Wrapf(err, "failed to find PV %v", genPVName(store))
+	} else if err = osc.checkPV(pv); err != nil {
+		logrus.Errorf("Object Store running but PV not ready")
+		return err
+	}
+
+	pvc, err := osc.ds.GetPersistentVolumeClaim(osc.namespace, genPVCName(store))
+	if err != nil && datastore.ErrorIsNotFound(err) {
+		return errors.Wrapf(err, "failed to find pvc %v", genPVCName(store))
+	} else if err = osc.checkPVC(pvc); err != nil {
+		logrus.Errorf("Object Store running but PVC not bound")
+		return err
+	}
+
+	return nil
+}
+
+func (osc *ObjectStoreController) handleStopping(store *longhorn.ObjectStore) (err error) {
+	dpl, err := osc.ds.GetDeployment(store.Name)
+	if err != nil {
+		return errors.Wrap(err, "failed find Deployment to stop")
+	} else if (*dpl).Spec.Replicas != nil && *((*dpl).Spec.Replicas) != 0 {
+		(*dpl).Spec.Replicas = int32Ptr(0)
+		_, err = osc.ds.UpdateDeployment(dpl)
+		return err
+	} else if dpl.Status.AvailableReplicas > 0 {
+		return nil // wait for shutdown
+	}
+
+	logrus.Infof("object store %v is now stopped", store.Name)
+	store.Status.State = longhorn.ObjectStoreStateStopped
+	return nil
+}
+
+func (osc *ObjectStoreController) handleStopped(store *longhorn.ObjectStore) (err error) {
+	if store.Spec.TargetState == longhorn.ObjectStoreStateRunning {
+		logrus.Infof("object store %v is now starting", store.Name)
+		store.Status.State = longhorn.ObjectStoreStateStarting
+		return nil
+	}
+	return nil
+}
+
+func (osc *ObjectStoreController) handleTerminating(store *longhorn.ObjectStore) (err error) {
```

Comment:
When the users delete an ObjectStore, can we make sure that ObjectStore controller will make sure to list all secrets with the label longhorn.io/object-store: <object-store-name> and longhorn.io/managed-by: longhorn-manager and delete them before removing the finalizer for the ObjectStore. This is the last part of the design https://github.com/longhorn/longhorn-manager/pull/2136#discussion_r1393442342 as we need to make sure we cleanup what we created